v1/cloudevents/transport/pubsub/internal/connection_test.go

```diff
@@ -63,26 +153,258 @@ func TestPublishCreateTopic(t *testing.T) {
+		})
 	}
+}
 
-	if err := psconn.DeleteTopic(ctx); err != nil {
-		t.Errorf("delete topic failed: %v", err)
+// Make sure that Publishing works if the original publish failed due to an
+// error in one of the pubsub calls.
+func TestPublishAfterPublishFailure(t *testing.T) {
+	for _, failureMethod := range []string{
+		"/google.pubsub.v1.Publisher/GetTopic",
+		"/google.pubsub.v1.Publisher/CreateTopic",
+		"/google.pubsub.v1.Publisher/Publish"} {
+		t.Run(failureMethod, func(t *testing.T) {
+			ctx := context.Background()
+			pc := &testPubsubClient{}
+			defer pc.Close()
+
+			projectID, topicID, subID := "test-project", "test-topic", "test-sub"
+
+			failureMap := make(map[string][]failPattern)
+			failureMap[failureMethod] = []failPattern{{
+				ErrReturn: fmt.Errorf("Injected error"),
+				Count:     1,
+				Delay:     0}}
+			client, err := pc.New(ctx, projectID, failureMap)
+			if err != nil {
+				t.Fatalf("failed to create pubsub client: %v", err)
+			}
+			defer client.Close()
+
+			psconn := &Connection{
+				AllowCreateSubscription: true,
+				AllowCreateTopic:        true,
+				Client:                  client,
+				ProjectID:               projectID,
+				TopicID:                 topicID,
+				SubscriptionID:          subID,
+			}
+
+			msg := &pubsub.Message{
+				ID:   "msg-id-1",
+				Data: []byte("msg-data-1"),
+			}
+			// Fails due to injected failure
+			if _, err := psconn.Publish(ctx, msg); err == nil {
+				t.Errorf("Expected publish failure, but didn't see it: %v", err)
+			}
+			// Succeeds
+			if _, err := psconn.Publish(ctx, msg); err != nil {
+				t.Errorf("failed to publish message: %v", err)
+			}
+			verifyTopicDeleteWorks(t, client, psconn, topicID)
+		})
 	}
+}
 
-	if ok, err := client.Topic(topicID).Exists(ctx); err != nil || ok {
+// Test Publishing after Deleting a first version of a topic
+func TestPublishCreateTopicAfterDelete(t *testing.T) {
+	ctx := context.Background()
+	pc := &testPubsubClient{}
+	defer pc.Close()
+
+	projectID, topicID, subID := "test-project", "test-topic", "test-sub"
+
+	client, err := pc.New(ctx, projectID, nil)
+	if err != nil {
+		t.Fatalf("failed to create pubsub client: %v", err)
+	}
+	defer client.Close()
+
+	psconn := &Connection{
+		AllowCreateSubscription: true,
+		AllowCreateTopic:        true,
+		Client:                  client,
+		ProjectID:               projectID,
+		TopicID:                 topicID,
+		SubscriptionID:          subID,
+	}
+
+	msg := &pubsub.Message{
+		ID:   "msg-id-1",
+		Data: []byte("msg-data-1"),
+	}
+	if _, err := psconn.Publish(ctx, msg); err != nil {
+		t.Errorf("failed to publish message: %v", err)
+	}
+
+	verifyTopicDeleteWorks(t, client, psconn, topicID)
+
+	if _, err := psconn.Publish(ctx, msg); err != nil {
+		t.Errorf("failed to publish message: %v", err)
+	}
+
+	verifyTopicDeleteWorks(t, client, psconn, topicID)
+}
+
+// Test that publishing fails if a toic doesn't exist and topic creation isn't allowed
```

Comment:
The reason for multiple pwople 

> Not sure under what context we're using topic/sub creation in this lib. It seems in knative-gcp we never really use this lib to create topic/sub. I'm curious if we can keep this lib simple by 1) removing support of multiple topics/subs 2) removing support of topic/sub creation.
> 
> Other than the general concern of the complexity here, this PR lgtm.

The reason for multiple openers of topics is that each event send call independently opens the topic if it isn't already open.  IE, it's a model that you configure the system for who it logically targets, then each send is responsible for opening a conneciton, reusing a conneciton, etc etc as needed.

Reply:
I understood the functionality (allow the sdk safely create conns and reuse them). I was curious about where this functionality is used? It seems pretty common we just manage topic/sub outside of the sdk.

Reply:
@slinkydeveloper @yolocs I think I agree that it may be better to try to move the topic management out of the internals of the API (or at least make each Sender its own topic and have the topic management in the creation and closing of each Sender)...   My suggestion would be that we only do that in v2 (I’d rather treat v1 is frozen except bug fixes) and preferably I open a separate Issue/PR as a follow-on rather than the pushing an interface change into a bug fix PR.  Thoughts?

Reply:
Sgtm!